[git-webkit revert] Add relationships between issues when reverting

briannafan opened the pull request:
#### e177e893ff985e413c4991ebe1d79c2a4f704c5c
<pre>
[git-webkit revert] Add relationships between issues when reverting
<a href="https://bugs.webkit.org/show_bug.cgi?id=267276">https://bugs.webkit.org/show_bug.cgi?id=267276</a>
<a href="https://rdar.apple.com/120723182">rdar://120723182</a>

Reviewed by Jonathan Bedard.

Adds &apos;blocks&apos; relationship between bugs and &apos;cause of&apos; relationship between radars.
This happens after reverting so the issues are only changed with a successful revert.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert.get_commit_info):
(Revert.get_issue_info):
(Revert):
(Revert.relate_issues):
(Revert.main):

Canonical link: <a href="https://commits.webkit.org/275111@main">https://commits.webkit.org/275111@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/846330929f067dfe320bcb6107bae6bfefc2afe6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40598 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19611 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42976 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43151 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36687 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42905 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22572 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16942 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33683 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41172 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16550 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34999 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14268 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/40468 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14343 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35972 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44424 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36812 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36300 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40039 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15413 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12673 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38370 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/40645 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17032 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9187 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17083 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16676 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -77,16 +77,24 @@ def get_commit_info(cls, args, repository, **kwargs):
+                elif isinstance(i.tracker, radar.Tracker):
+                    commit_radars.add(i)
+
+        return commit_objects, commits_to_revert, commit_bugs, commit_radars
```
I do wonder about this API, though I realise it's essentially just extending what's already here.

`commit_objects` already actually contains everything, so everything else in the tuple is effectively duplicating it.

Especially with a single caller, it feels like it would be just as easy to have the caller just do things like `commits_to_revert = [c.hash for c in commit_objects]`?

briannafan replied:
That's a good point about the hashes! The bugs and radars will be used multiple times so I'd like to keep those.

gsnedders reviewed (commented):
(a singular non-reviewer comment; the rest seems fine to me, and this is really a pre-existing nit)

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -75,18 +75,24 @@ def parser(cls, parser, loggers=None):
     def get_commit_info(cls, args, repository, **kwargs):
         commit_objects = []
-        commits_to_revert = []
         commit_bugs = set()
```
Feels like we can generalize this code pretty easily.

In general, this isn't creating separate lists for Bugzillas and radars, it's creating a separate list for each tracker we have. So `commit_bugs = set()` should really be `commit_issues = {}` where each key represents a tracker (perhaps `Tracker.name` as discussed in #24026)

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -75,18 +75,24 @@ def parser(cls, parser, loggers=None):
-            commit_bugs.update({Tracker.from_string(i.link) for i in commit.issues if isinstance(Tracker.from_string(i.link).tracker, bugzilla.Tracker)})
-        return commit_objects, commits_to_revert, commit_bugs
+            for i in commit.issues:
+                if isinstance(i.tracker, bugzilla.Tracker):
```
This should probably be something like:
```
commit_issues[i.tracker.name].add(i)
```

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -227,10 +235,42 @@ def revert_commit(cls, args, repository, issue, commit_objects, commit_bugs, com
+        return 0
+
+    @classmethod
+    def relate_issues(cls, args, repository, issue, commit_bugs, commit_radars, revert_reason):
```
Looking through this function, the only bits that are unique to specific trackers are `cc_radar`, catching the `UnsuccessfulResponseException` and `blocks` vs `cause_of`.

In general, you basically have two buckets. The original issues (in all bug trackers) and then the revert issues (also in all bug trackers). You're trying to re-open all original issues and relate them to the paired revert issue in their specific bug tracker. I would like to see if we could use unified logic until the very last moment, and ask ourselves if there is any way to unify even that last moment (there may not be, I don't think `cc_radar` has an idiom that makes sense when invoked on a radar, for example)

briannafan replied:
That's a great summary! I tried implementing it so we only need to condition on tracker type when relating as you mentioned. I checked and `cc_radar` will return `None` if called on a radar which is nice.

briannafan committed:
[git-webkit revert] Add relationships between issues when reverting
https://bugs.webkit.org/show_bug.cgi?id=267276
rdar://120723182

Reviewed by Jonathan Bedard.

Adds 'blocks' relationship between bugs and 'cause of' relationship between radars.
This happens after reverting so the issues are only changed with a successful revert.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert.get_commit_info):
(Revert.get_issue_info):
(Revert):
(Revert.relate_issues):
(Revert.main):

Canonical link: https://commits.webkit.org/275111@main
